### PR TITLE
Run the JavaScript kernel in a Web Worker

### PR DESCRIPTION
**Pull request by jtpio**
## References

More steps towards https://github.com/jupyterlite/jupyterlite/issues/463 to streamline how the JupyterLite kernels are set up.

## Code changes

- [x] Run the JavaScript kernel in a Web Worker instead of an IFrame
- [x] Some minor tweaks and docstrings on the Pyolite kernel
- [x] Check downstream kernels like p5 -> Will handle updates in https://github.com/jupyterlite/p5-kernel/pull/11

## User-facing changes

TBD

## Backwards-incompatible changes

Some changes in the JavaScript kernel:

- remove `protected _eval()` method

### CODE REVIEW

**Comment by jtpio** [comment]
> * Check downstream kernels like p5 -> Will handle updates in

Changes will be required downsteam, but the p5 kernel will first need the new `@jupyterlite` packages to be released.

File: packages/javascript-kernel/src/kernel.ts [hunk 1]
```diff
@@ -233,61 +177,108 @@ export class JavaScriptKernel extends BaseKernel implements IKernel {
...
-        console.warn = console.error;
-
-        window.onerror = function(message, source, lineno, colno, error) {
-          console.error(message);
-        }
-      `
-    );
+    switch (msg.type) {
+      case 'stream': {
+        const bundle = msg.bundle ?? { name: 'stdout', text: '' };
```
Comment (bollwyvl) [inline packages/javascript-kernel/src/kernel.ts:221]: maybe all the empty responses could be hoisted to  `emptyStreamResponse = Object.freeze({name: 'stdout': text: ''})`, e.g constants up top?

File: packages/pyolite-kernel/src/worker.ts [hunk 2]
```diff
@@ -326,16 +326,13 @@ export class PyoliteRemoteKernel {
    *
    * @param content The incoming message with the comm target name.
    */
-  async commInfo(content: any, parent: any) {
+  async commInfo(content: { target_name: string }, parent: any) {
```
Comment (bollwyvl) [inline packages/pyolite-kernel/src/worker.ts:329]: seems like we should hoist that `content` to a type.

File: packages/javascript-kernel/src/kernel.ts [hunk 3]
```diff
@@ -15,25 +21,10 @@ export class JavaScriptKernel extends BaseKernel implements IKernel {
...
-      window.addEventListener('message', (e: MessageEvent) => {
-        const msg = e.data;
-        if (msg.event === 'stream') {
-          const content = msg as KernelMessage.IStreamMsg['content'];
-          this.stream(content);
-        }
-      });
-    };
-    document.body.appendChild(this._iframe);
+    this._worker = this.initWorker();
```
Comment (bollwyvl) [inline packages/javascript-kernel/src/kernel.ts:24]: Same thing here: i think passing the options down is a good idea for "base" kernels like this.

Comment (jtpio) [reply packages/javascript-kernel/src/kernel.ts:24]: Good point :+1:

File: packages/pyolite-kernel/src/kernel.ts [hunk 4]
```diff
@@ -36,12 +36,18 @@ export class PyoliteKernel extends BaseKernel implements IKernel {
    * Subclasses must implement this typographically almost _exactly_ for
    * webpack to find it.
    */
-  protected initWorker(options: PyoliteKernel.IOptions): Worker {
+  protected initWorker(): Worker {
```
[the remark this answers is not in the document]
Comment (jtpio) [reply packages/pyolite-kernel/src/kernel.ts:39]: Makes sense based on the comment above: https://github.com/jupyterlite/jupyterlite/pull/711#pullrequestreview-1213706627

File: packages/pyolite-kernel/src/worker.ts (hunk 2 above)
Comment (jtpio) [reply packages/pyolite-kernel/src/worker.ts:329]: Might revert back to `any` for now and do a pass on the types separately on the other methods as well.

**Comment by jtpio** [comment]
Getting this one in and following-up with a new release right after.

**Pull request merged**